src/main/scala/com/databricks/spark/csv/package.scala

```diff
@@ -45,28 +54,29 @@ package object csv {
   implicit class CsvSchemaRDD(dataFrame: DataFrame) {
     def saveAsCsvFile(path: String, parameters: Map[String, String] = Map()): Unit = {
       // TODO(hossein): For nested types, we may want to perform special work
-      val delimiter = parameters.getOrElse("delimiter", ",")
+      val delimiter = parameters.getOrElse("delimiter", ",").charAt(0)
+      val quote = parameters.getOrElse("quote", "\"").charAt(0)
+      val escape = parameters.getOrElse("escape", "\\").charAt(0)
       val generateHeader = parameters.getOrElse("header", "false").toBoolean
-      val header = if (generateHeader) {
-        dataFrame.columns.map(c => s""""$c"""").mkString(delimiter)
-      } else {
-        "" // There is no need to generate header in this case
-      }
-      val strRDD = dataFrame.rdd.mapPartitions { iter =>
-        new Iterator[String] {
-          var firstRow: Boolean = generateHeader
+      val header = dataFrame.columns
 
-          override def hasNext = iter.hasNext
+      var firstRow: Boolean = generateHeader
+      val csvFileFormat = CSVFormat.DEFAULT
+      .withDelimiter(delimiter)
+      .withQuote(quote)
+      .withEscape(escape)
 
-          override def next: String = {
-            if (firstRow) {
-              firstRow = false
-              header + "\n" + iter.next.mkString(delimiter)
-            } else {
-              iter.next.mkString(delimiter)
-            }
-          }
+      val strRDD = dataFrame.rdd.mapPartitions { iter =>
+        var firstRow: Boolean = generateHeader
+        val newIter = iter.map(_.toSeq.toArray)
+        val stringWriter = new StringWriter()
+        val csvPrinter = new CSVPrinter(stringWriter, csvFileFormat)
+        if (firstRow) {
+          firstRow = false
+          csvPrinter.printRecord(header:_*)
         }
+        csvPrinter.printRecords(WrapAsJava.asJavaIterable(newIter.toIterable))
```

Comment:
This can be inefficient because it seems to me we are traversing the iterator multiple times for each partition. I know CSVPrinter API is not very flexible here, but can we avoid multiple traversals?